app/src/main/java/org/luxoft/sdl_core/BluetoothLongReader.java

```diff
@@ -27,38 +39,64 @@ public void setCallback(LongReaderCallback callback) {
+        final int frames_count = header_bytes.getInt();
+        final boolean is_compressed = header_bytes.get() == 1;
+        return new MessageInfo(frames_count, is_compressed);
     }
 
     private byte[] extractMessage(byte[] value) {
         return Arrays.copyOfRange(value, mMessageOffset, value.length);
     }
 
     public void processReadOperation(byte[] value) {
-        int consecutive_frames = extractFramesCount(value); // first 4 bytes always contains amount of frames left
+        MessageInfo m_info = extractMessageInfo(value);
 
         if (mBuffer == null) {
-            Log.d(TAG, "Allocating new buffer size for " + consecutive_frames);
-            mBuffer = ByteBuffer.allocate(mMtu * (consecutive_frames + 1));
+            Log.d(TAG, "Allocating new buffer size for " + m_info.frames_count);
+            mBuffer = ByteBuffer.allocate(mMtu * (m_info.frames_count + 1));
             Log.d(TAG, "New buffer size is " + mBuffer.capacity());
         }
 
         byte[] message = extractMessage(value);
         mBuffer.put(message);
 
-        if (consecutive_frames > 0) {
+        if (m_info.frames_count > 0) {
             Log.d(TAG, "Added message of size " + message.length + " and remain " +
-                    mBuffer.remaining() + " bytes and " + consecutive_frames + " frames");
+                    mBuffer.remaining() + " bytes and " + m_info.frames_count + " frames " +
+                    " compression :" + m_info.is_compressed);
         } else {
-            Log.d(TAG, "Added last message of size " + message.length + " - passing to handler " + (mBuffer.capacity() - mBuffer.remaining()));
+            final int msg_size = mBuffer.capacity() - mBuffer.remaining();
+            Log.d(TAG, "Added last message of size " + message.length +
+                    " - passing to handler " + msg_size +
+                    " compression :" + m_info.is_compressed);
             if (mCallback != null) {
-                final byte[] full_message = Arrays.copyOf(mBuffer.array(), mBuffer.capacity() - mBuffer.remaining());
+                byte[] full_message = getFullMessage(m_info.is_compressed, msg_size);
                 mCallback.OnLongMessageReceived(full_message);
             }
             mBuffer.clear();
             mBuffer = null;
         }
     }
+
+    private byte[] getFullMessage(final boolean is_compressed, final int msg_size) {
+        final byte[] full_message_ = Arrays.copyOf(mBuffer.array(), msg_size);
+        byte[] full_message = full_message_;
+
+        if(is_compressed) {
+            try {
+                full_message = CompressionUtil.decompress(full_message_);
+            } catch (IOException e) {
+                full_message = full_message_;
+                e.printStackTrace();
+            } catch (DataFormatException e) {
+                full_message = full_message_;
+                e.printStackTrace();
+            }
+        }
+
+        return full_message;
```

Comment:
```suggestion
        final byte[] full_message = Arrays.copyOf(mBuffer.array(), msg_size);

        if(is_compressed) {
            try {
                return CompressionUtil.decompress(full_message);
            } catch (IOException e) {
                e.printStackTrace();
            } catch (DataFormatException e) {
                e.printStackTrace();
            }
        }

        return full_message;
```

Reply:
done in 568f87bcadda8f2ec653f1aa5a8651fa787e7d8c